### PR TITLE
CORE-7647 Refactor VirtualNodeDb to support testing

### PR DESCRIPTION
**Pull request by owenstanford**
Remove some of the tight coupling in the code that makes it hard to write new tests for the create node async change.

### CODE REVIEW

File: components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/VirtualNodeDbFactoryImpl.kt [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+import net.corda.db.core.DbPrivilege.DDL
+import net.corda.db.core.DbPrivilege.DML
+import net.corda.schema.configuration.DatabaseConfig
+import net.corda.virtualnode.ShortHash
+import java.security.SecureRandom
+
+/**
+ * A factory for [VirtualNodeDb]s.
+ */
+class VirtualNodeDbFactoryImpl(
```
Comment (owenstanford) [inline components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/VirtualNodeDbFactoryImpl.kt]: This is the original unmodified class with a new interface, the renaming has not been picked up by git as a move

File: components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/VirtualNodeDbImpl.kt [hunk 2]
```diff
@@ -0,0 +1,116 @@
+package net.corda.virtualnode.write.db.impl.writer
```
Comment (owenstanford) [inline components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/VirtualNodeDbImpl.kt:1]: This is the original class with a new interface, the renaming has not been picked up by git as a move

File: components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/DbConnectionImpl.kt [hunk 3]
```diff
@@ -0,0 +1,21 @@
+package net.corda.virtualnode.write.db.impl.writer
```
Comment (owenstanford) [inline components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/DbConnectionImpl.kt:1]: This is the original class with a new interface, the renaming has not been picked up by git as a move

File: components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/DbConnection.kt [hunk 4]
```diff
@@ -2,30 +2,31 @@ package net.corda.virtualnode.write.db.impl.writer
...
+ * Represents a Virtual Node DB connection
  *
  * @property name Connection name
  * @property privilege Connection's DB privilege
  * @property config Connection configuration
  * @property description Connection's description
  */
-class DbConnection(val name: String, val privilege: DbPrivilege, val config: SmartConfig, val description: String) {
+
+interface DbConnection {
```
Comment (vkolomeyko) [inline components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/DbConnection.kt]:
```suggestion
internal interface DbConnection {
```
just to highlight the fact that it is not going to be used outside of this module.

File: components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/VirtualNodeDb.kt [hunk 5]
```diff
@@ -1,98 +1,34 @@
...
-class VirtualNodeDb(
-    val dbType: VirtualNodeDbType, val isClusterDb: Boolean, private val holdingIdentityShortHash: ShortHash,
-    val dbConnections: Map<DbPrivilege, DbConnection?>, private val dbAdmin: DbAdmin,
-    private val dbConnectionManager: DbConnectionManager, private val schemaMigrator: LiquibaseSchemaMigrator
-) {
-
-    companion object {
-        private val log = LoggerFactory.getLogger(this::class.java.enclosingClass)
-    }
+interface VirtualNodeDb {
```
Comment (vkolomeyko) [inline components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/VirtualNodeDb.kt]:
```suggestion
internal interface VirtualNodeDb {
```

File: components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/VirtualNodeDbFactory.kt [hunk 6]
```diff
@@ -1,41 +1,10 @@
...
-
-    companion object {
-        private const val ddlMaxPoolSize = 1
-        private const val dmlMaxPoolSize = 1
-        private const val passwordLength = 64
-        private val passwordSource = (('0'..'9') + ('A'..'Z') + ('a'..'z')).toCharArray()
-        private val random = SecureRandom()
-    }
 
+interface VirtualNodeDbFactory {
```
Comment (vkolomeyko) [inline components/virtual-node/virtual-node-write-service-impl/src/main/kotlin/net/corda/virtualnode/write/db/impl/writer/VirtualNodeDbFactory.kt]:
```suggestion
internal interface VirtualNodeDbFactory {
```

**Review by vkolomeyko — COMMENTED** [review]
Looks fine.
Minor comments/suggestions posted.

**Review by vkolomeyko — APPROVED** [review]
LGTM

**Pull request merged**